File: packages/ogdf.js/src/layouts/module/InitialPlacer.js
import createModule from "./module-helper";
import { PARAMETER_TYPE } from "../../utils/parameter-type";
const NAME = "InitialPlacer"

const MODULE_DIRECTORY = {
    BarycenterPlacer: {
        randomOffset: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        weightedPositionPriority: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        }
    },
    CirclePlacer: {
        circleSize: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 0.0
        },
        nodeSelection: {
            type: PARAMETER_TYPE.CATEGORICAL,
            range: ['New', 'Old', 'All'],
            default: 'New'
        },
        radiusFixed: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        randomOffset: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        }
    },
    MedianPlacer: {
        randomOffset: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        }
    },
    RandomPlacer: {
        randomOffset: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        circleSize: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 0.0
        }
    },
    SolarPlacer: {
        randomOffset: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        }
    },
    ZeroPlacer: {
        randomRange: {
            type: PARAMETER_TYPE.DOUBLE,
            range: [0, Infinity],
            default: 1.0
        },
        randomOffset: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        }
    }
}

const InitialPlacer = createModule(NAME, MODULE_DIRECTORY)
export default InitialPlacer

File: packages/ogdf.js/src/module/src/InitialPlacer.cpp
#include "../main.h"

#include <ogdf/energybased/multilevel_mixer/BarycenterPlacer.h>
#include <ogdf/energybased/multilevel_mixer/CirclePlacer.h>
#include <ogdf/energybased/multilevel_mixer/MedianPlacer.h>
#include <ogdf/energybased/multilevel_mixer/RandomPlacer.h>
#include <ogdf/energybased/multilevel_mixer/SolarPlacer.h>
#include <ogdf/energybased/multilevel_mixer/ZeroPlacer.h>

EM_PORT_API(BarycenterPlacer *)
InitialPlacer_BarycenterPlacer(bool randomOffset, bool weightedPositionPriority)
{
    BarycenterPlacer *placer = new BarycenterPlacer();
    placer->setRandomOffset(randomOffset);
    placer->weightedPositionPriority(weightedPositionPriority);
    return placer;
}

EM_PORT_API(CirclePlacer *)
InitialPlacer_CirclePlacer(float circleSize, int nodeSelection, bool radiusFixed, bool randomOffset)
{
    CirclePlacer *placer = new CirclePlacer();
    placer->setCircleSize(circleSize);
    placer->setNodeSelection(static_cast<CirclePlacer::NodeSelection>(nodeSelection));
    placer->setRadiusFixed(radiusFixed);
    placer->setRandomOffset(randomOffset);
    return placer;
}

EM_PORT_API(MedianPlacer *)
InitialPlacer_MedianPlacer(bool randomOffset)
{
    MedianPlacer *placer = new MedianPlacer();
    placer->setRandomOffset(randomOffset);
    return placer;
}

EM_PORT_API(RandomPlacer *)
InitialPlacer_RandomPlacer(bool randomOffset, double circleSize)
{
    RandomPlacer *placer = new RandomPlacer();
    placer->setRandomOffset(randomOffset);
    placer->setCircleSize(circleSize);
    return placer;
}

EM_PORT_API(SolarPlacer *)
InitialPlacer_SolarPlacer(bool randomOffset)
{
    SolarPlacer *placer = new SolarPlacer();
    placer->setRandomOffset(randomOffset);
    return placer;
}

EM_PORT_API(ZeroPlacer *)
InitialPlacer_ZeroPlacer(double randomRange, bool randomOffset)
{
    ZeroPlacer *placer = new ZeroPlacer();
    placer->setRandomRange(randomRange);
    placer->setRandomOffset(randomOffset);
    return placer;
}
